pkg/sdn/plugin/node_iptables.go

```diff
@@ -165,23 +152,53 @@ func (n *NodeIPTables) getNodeIPTablesChains() []Chain {
 				{"-i", "docker0", "-m", "comment", "--comment", "from docker to localhost", "-j", "ACCEPT"},
 			},
 		},
-		{
-			table:    "filter",
-			name:     "OPENSHIFT-FIREWALL-FORWARD",
-			srcChain: "FORWARD",
-			srcRule:  []string{"-m", "comment", "--comment", "firewall overrides"},
-			rules: [][]string{
-				{"-s", n.clusterNetworkCIDR, "-m", "comment", "--comment", "attempted resend after connection close", "-m", "conntrack", "--ctstate", "INVALID", "-j", "DROP"},
-				{"-d", n.clusterNetworkCIDR, "-m", "comment", "--comment", "forward traffic from SDN", "-j", "ACCEPT"},
-				{"-s", n.clusterNetworkCIDR, "-m", "comment", "--comment", "forward traffic to SDN", "-j", "ACCEPT"},
-			},
-		},
-		{
+		Chain{
 			table:    "filter",
 			name:     "OPENSHIFT-ADMIN-OUTPUT-RULES",
 			srcChain: "FORWARD",
 			srcRule:  []string{"-i", Tun0, "!", "-o", Tun0, "-m", "comment", "--comment", "administrator overrides"},
 			rules:    nil,
-		},
+		})
+
+	for _, cidr := range n.clusterNetworkCIDR {
+		var masqRule []string
+
+		if n.masqueradeServices {
+			masqRule = []string{"-s", cidr, "-m", "comment", "--comment", "masquerade pod-to-service and pod-to-external traffic", "-j", "MASQUERADE"}
+		} else {
+			masqRule = []string{"-s", cidr, "!", "-d", cidr, "-m", "comment", "--comment", "masquerade pod-to-external traffic", "-j", "MASQUERADE"}
```

Comment:
This won't work right. The rule is supposed to be "if a packet starts in the SDN and doesn't go back into the SDN, it needs to be masqueraded". But eg, if you had two networks 1.2.3.0/8 and 4.5.6.0/8 then this would generate:

    -A OPENSHIFT-MASQUERADE -s 1.2.3.0/8 ! -d 1.2.3.0/8 -j MASQUERADE
    -A OPENSHIFT-MASQUERADE -s 4.5.6.0/8 ! -d 4.5.6.0/8 -j MASQUERADE

and so if a packet went from 1.2.3.4 to 4.5.6.7, it would end up getting masqueraded (which you didn't want) because it matches the first rule.

This isn't totally trivial to fix. One way would be to have a second table; in OPENSHIFT-MASQUERADE you'd have a series of rules to match packets whose source was in the SDN:

    -A OPENSHIFT-MASQUERADE -s 1.2.3.0/8 -j OPENSHIFT-MASQUERADE-2
    -A OPENSHIFT-MASQUERADE -s 4.5.6.0/8 -j OPENSHIFT-MASQUERADE-2

and then in OPENSHIFT-MASQUERADE-2 you'd bail out if the destination was also in the SDN, but masquerade otherwise:

    -A OPENSHIFT-MASQUERADE-2 -d 1.2.3.0/8 -j RETURN
    -A OPENSHIFT-MASQUERADE-2 -d 4.5.6.0/8 -j RETURN
    -A OPENSHIFT-MASQUERADE-2 -j MASQUERADE

It's totally possible there's a simpler way to do this; I'm not super-duper iptables-savvy. @dcbw might have ideas?

Reply:
@danwinship @JacobTanenbaum I'm probably missing something here, but shouldn't packets that are CIDR->CIDR stay in the SDN and never get to tun0 in the first place?  If a packet gets to tun0 that is meant directly for another pod in the SDN, then our OVS rules have failed and that packet should probably get dropped.

eg, in the above example, I can't see why we'd ever get a packet from 1.2.3.4 -> 4.5.6.7 hitting tun0 and thus iptables.  That should just get handled via OVS rules and send directly to the other node in the VXLAN, not outside the SDN.

Or am I missing something in architecture here that I should have seen...?

Reply:
> I'm probably missing something here, but shouldn't packets that are CIDR->CIDR stay in the SDN and never get to tun0 in the first place?

Yes, you're missing something (though it's totally non-obvious so we probably should have a comment here). This rule applies to packets that go pod -> service IP -> pod; after the iptables service DNAT rule applies, it will have source and destination both in clusterNetworkCIDR(s), but it needs to be masqueraded to ensure that it goes back through iptables again on the way back (to get un-DNATed)